src/uct/cuda/cuda_copy/cuda_copy_iface.c

```diff
@@ -86,27 +86,116 @@ static ucs_status_t uct_cuda_copy_iface_query(uct_iface_h iface,
     return UCS_OK;
 }
 
+static ucs_status_t uct_cuda_copy_iface_flush(uct_iface_h tl_iface, unsigned flags,
+                                              uct_completion_t *comp)
+{
+    uct_cuda_copy_iface_t *iface = ucs_derived_of(tl_iface, uct_cuda_copy_iface_t);
+
+    if (comp != NULL) {
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    if (ucs_queue_is_empty(&iface->outstanding_d2h_cuda_event_q) &&
+        ucs_queue_is_empty(&iface->outstanding_h2d_cuda_event_q)) {
+        UCT_TL_IFACE_STAT_FLUSH(ucs_derived_of(tl_iface, uct_base_iface_t));
+        return UCS_OK;
+    }
+
+    UCT_TL_IFACE_STAT_FLUSH_WAIT(ucs_derived_of(tl_iface, uct_base_iface_t));
+    return UCS_INPROGRESS;
+}
+
+static UCS_F_ALWAYS_INLINE unsigned
+uct_cuda_copy_progress_event_queue(ucs_queue_head_t *event_queue)
+{
+    unsigned count = 0;
+    cudaError_t result = cudaSuccess;
+    uct_cuda_copy_event_desc_t *cuda_event;
+    ucs_queue_iter_t iter;
+
+    ucs_queue_for_each_safe(cuda_event, iter, event_queue, queue) {
+        result = cudaEventQuery(cuda_event->event);
+        if (cudaSuccess != result) {
+            break;
+        }
+        ucs_queue_del_iter(event_queue, iter);
+        if (cuda_event->comp != NULL) {
+            uct_invoke_completion(cuda_event->comp, UCS_OK);
+        }
+        ucs_info("CUDA Event Done :%p", cuda_event);
```

Comment:
probably it should be ucx_trace_poll() ?

Reply:
fixed